test/hotspot/jtreg/compiler/loopopts/superword/TestMulAddS2I.java

```diff
@@ -0,0 +1,96 @@
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/**
+ * @test
+ * @bug 8310886
+ * @requires os.arch == "x86_64" | os.arch == "aarch64"
+ * @summary Test MulAddS2I vectorization.
+ * @library /test/lib /
+ * @run driver compiler.loopopts.superword.TestMulAddS2I
+ */
+
+package compiler.loopopts.superword;
+
+import compiler.lib.ir_framework.*;
+import jdk.test.lib.Asserts;
+import jdk.test.lib.Platform;
+
+public class TestMulAddS2I {
+    static final int RANGE = 1024;
+    static final int ITER  = RANGE/2 - 1;
+
+    static short[] sArr1 = new short[RANGE];
+    static short[] sArr2 = new short[RANGE];
+    static final int[] GOLDEN;
+
+    static {
+        for (int i = 0; i < RANGE; i++) {
+            sArr1[i] = (short)(i + (1000 * AbstractInfo.getRandom().nextDouble()));
+            sArr2[i] = (short)(RANGE - i + (2000 * AbstractInfo.getRandom().nextDouble()));
+        }
+        GOLDEN = test();
+    }
+
+
+    public static void main(String[] args) {
+        if (Platform.isX64() || Platform.isX86()) {
+            TestFramework.runWithFlags("-XX:+UseUnalignedLoadStores");
+            TestFramework.runWithFlags("-XX:-UseUnalignedLoadStores");
+        } else {
+            TestFramework.run();
+        }
+    }
+
+    @Run(test = "test")
+    @Warmup(0)
+    public static void run() {
+        compare(test());
+    }
+
+    public static void compare(int[] out) {
+        for (int i = 0; i < ITER; i++) {
+            Asserts.assertEQ(out[i], GOLDEN[i], "wrong result for out[" + i + "]");
+        }
+    }
+
+    @Test
+    @IR(applyIfCPUFeature = {"sse2", "true"}, applyIf = {"UseUnalignedLoadStores", "true"},
+        counts = {IRNode.MUL_ADD_S2I, "> 0", IRNode.MUL_ADD_VS2VI, "> 0"})
+    @IR(applyIfCPUFeature = {"sse2", "true"}, applyIf = {"UseUnalignedLoadStores", "false"},
+        failOn = {IRNode.MUL_ADD_VS2VI}, // Can only pack LoadS if UseUnalignedLoadStores is true (default if sse4.2)
+        counts = {IRNode.MUL_ADD_S2I, "> 0"})
+    @IR(applyIfCPUFeature = {"asimd", "true"}, applyIf = {"MaxVectorSize", "16"},
```

Comment:
Why `"MaxVectorSize", "16"`?

Reply:
The AD file enforces it:
https://github.com/openjdk/jdk/blob/5350fd617390aaaedf8dd8821418c796cb1c38b3/src/hotspot/cpu/aarch64/aarch64_vector.ad#L182-L185

I'm not sure if that limitation is too restrictive. I could be investigated in a separate RFE. But I think to ensure the correctness of this rule, we need this constraint for aarch64.